File: python_implementation/mergesort_top_down/mergesort_top_down.py
def mergesort_top_down(list):
    if (len(list) == 1):
        return list

    middle = len(list) / 2

    a = mergesort(list[0, middle])
    b = mergesort(list[middle, len(list)])
    ab = merge(a, b)
    return ab

def merge(a, b):
    result = []
    c = []

    while (len(a) != 0 and len(b) != 0):
        if (a[0] < b[0]):
            result.append(a[0])
            del a[0]
        else:
            result.append(b[0])
            del b[0]

        if (len(a) == 0):
            c = b
        elif (len(b) == 0):
            c = a

    for i in c:
        result.append(i)

    return result

File: java_implementation/array/mergesort_top_down/mergesort_top_down.java
package java_implementation.array.mergesort_top_down;

public class mergesort_top_down
{
  public int[] sort(int[] array)
  {
    if (array.length == 1)
    {
      return array;
    }

    int middle = array.length / 2;

    int acopy[] = new int[middle];
    System.arraycopy(array, 0, acopy, 0, acopy.length);
    int a[] = sort(acopy);

    int bcopy[] = new int[array.length - middle]; // -1
    System.arraycopy(array, middle, bcopy, 0, bcopy.length);
    int b[] = sort(bcopy);

    int ab[] = merge(a, b);
    return ab;
  }

  public int[] merge(int a[], int b[])
  {
    int result[] = new int[a.length + b.length];
    int index = 0;
    int i = 0, j = 0;

    while (i != a.length && j != b.length)
    {
      if (a[i] < b[j])
      {
        result[index] = a[i];
        index++;
        i++;
      }
      else
      {
        result[index] = b[j];
        index++;
        j++;
      }
    }

    if (i == a.length)
    {
      while (j < b.length)
      {
        result[index] = b[j];
        index++;
        j++;
      }
    }
    else if (j == b.length)
    {
      while (i < a.length)
      {
        result[index] = a[i];
        index++;
        i++;
      }
    }

    return result;
  }
}
